File: results/dnd-rest-benchmark/lifecycle-runs/20260731T051216Z_claude_claude-sonnet-5_php-stdlib/lib/play_safe_turns.php
<?php
declare(strict_types=1);

// Routes: Play campaign concurrent-safe turn submission
// ---------------------------------------------------------------------------

// Returns the current safe-turn counter for a campaign, creating the row
// (starting at turn 1) if this is the first submission/read for it.
function get_or_init_safe_turn_state(PDO $db, string $campaignId): int {
    $stmt = $db->prepare('SELECT current_turn FROM play_campaign_safe_turns WHERE campaign_id = ?');
    $stmt->execute([$campaignId]);
    $row = $stmt->fetch(PDO::FETCH_ASSOC);
    if ($row !== false) {
        return (int)$row['current_turn'];
    }
    $stmt = $db->prepare('INSERT OR IGNORE INTO play_campaign_safe_turns (campaign_id, current_turn) VALUES (?, 1)');
    $stmt->execute([$campaignId]);
    return 1;
}

if ($method === 'POST' && preg_match('#^/v1/play/campaigns/([^/]+)/safe-turns$#', $path, $pm)) {
    $campaignId = $pm[1];
    $actor = require_actor($db);

    $campaign = require_play_campaign($db, $campaignId);

    $isOwner = $actor['username'] === $campaign['owner'];
    if (!$isOwner && !is_campaign_member($db, $campaignId, $actor['username'])) {
        forbidden('must be a campaign member to submit safe turns');
    }

    $body = read_json_body();
    if ($body === null
        || !isset($body['submission_id'], $body['expected_turn'], $body['action'])
        || !is_string($body['submission_id']) || trim($body['submission_id']) === ''
        || !is_string($body['action']) || trim($body['action']) === ''
        || !is_valid_int_range($body['expected_turn'], 1, PHP_INT_MAX)) {
        bad_request('submission_id, action, and a positive integer expected_turn are required');
    }
    $submissionId = $body['submission_id'];
    $action = $body['action'];
    $expectedTurn = (int)$body['expected_turn'];

    $db->beginTransaction();
    try {
        $currentTurn = get_or_init_safe_turn_state($db, $campaignId);

        $stmt = $db->prepare('SELECT 1 FROM play_campaign_safe_turn_submissions WHERE campaign_id = ? AND submission_id = ?');
        $stmt->execute([$campaignId, $submissionId]);
        if ($stmt->fetch(PDO::FETCH_ASSOC) !== false) {
            $db->commit();
            conflict('submission_id already used in this campaign');
        }

        if ($expectedTurn !== $currentTurn) {
            $db->commit();
            send_json(['current_turn' => $currentTurn], 409);
        }

        $acceptedTurn = $currentTurn;
        $nextTurn = $currentTurn + 1;

        $stmt = $db->prepare('SELECT COALESCE(MAX(sequence), 0) AS max_seq FROM play_campaign_safe_turn_submissions WHERE campaign_id = ?');
        $stmt->execute([$campaignId]);
        $sequence = (int)$stmt->fetch(PDO::FETCH_ASSOC)['max_seq'] + 1;

        $stmt = $db->prepare('INSERT INTO play_campaign_safe_turn_submissions (campaign_id, sequence, submission_id, action, accepted_turn, next_turn) VALUES (?, ?, ?, ?, ?, ?)');
        $stmt->execute([$campaignId, $sequence, $submissionId, $action, $acceptedTurn, $nextTurn]);

        $stmt = $db->prepare('UPDATE play_campaign_safe_turns SET current_turn = ? WHERE campaign_id = ?');
        $stmt->execute([$nextTurn, $campaignId]);

        $db->commit();
    } catch (Throwable $e) {
        if ($db->inTransaction()) {
            $db->rollBack();
        }
        throw $e;
    }

    send_json([
        'submission_id' => $submissionId,
        'action' => $action,
        'accepted_turn' => $acceptedTurn,
        'next_turn' => $nextTurn,
    ], 201);
}

if ($method === 'GET' && preg_match('#^/v1/play/campaigns/([^/]+)/safe-turns$#', $path, $pm)) {
    $campaignId = $pm[1];
    $actor = require_actor($db);

    $campaign = require_play_campaign($db, $campaignId);

    $isOwner = $actor['username'] === $campaign['owner'];
    if (!$isOwner && !is_campaign_member($db, $campaignId, $actor['username'])) {
        forbidden('must be a campaign member to read safe turns');
    }

    $currentTurn = get_or_init_safe_turn_state($db, $campaignId);

    $stmt = $db->prepare('SELECT submission_id, action, accepted_turn, next_turn FROM play_campaign_safe_turn_submissions WHERE campaign_id = ? ORDER BY sequence ASC');
    $stmt->execute([$campaignId]);
    $rows = $stmt->fetchAll(PDO::FETCH_ASSOC);

    $accepted = array_map(function (array $row): array {
        return [
            'submission_id' => $row['submission_id'],
            'action' => $row['action'],
            'accepted_turn' => (int)$row['accepted_turn'],
            'next_turn' => (int)$row['next_turn'],
        ];
    }, $rows);

    send_json([
        'current_turn' => $currentTurn,
        'accepted' => $accepted,
    ]);
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260730T082348Z_claude_claude-sonnet-5_go-stdlib/play_safe_turns.go
package main

import (
	"encoding/json"
	"net/http"
)

// playSafeTurn is an accepted campaign-scoped safe turn submission.
type playSafeTurn struct {
	SubmissionID string `json:"submission_id"`
	Action       string `json:"action"`
	AcceptedTurn int    `json:"accepted_turn"`
	NextTurn     int    `json:"next_turn"`
}

// handlePlayCampaignSafeTurnsSub routes the "safe-turns" sub-path of a play
// campaign.
func handlePlayCampaignSafeTurnsSub(w http.ResponseWriter, r *http.Request, id, rest string) bool {
	if rest != "safe-turns" {
		return false
	}
	switch r.Method {
	case http.MethodPost:
		handleSubmitPlaySafeTurn(w, r, id)
	case http.MethodGet:
		handleListPlaySafeTurns(w, r, id)
	default:
		writeError(w, http.StatusMethodNotAllowed, "method not allowed")
	}
	return true
}

func handleSubmitPlaySafeTurn(w http.ResponseWriter, r *http.Request, campaignID string) {
	username, ok := authenticatePlay(r)
	if !ok {
		writeError(w, http.StatusUnauthorized, "authentication required")
		return
	}

	var req struct {
		SubmissionID string `json:"submission_id"`
		ExpectedTurn int    `json:"expected_turn"`
		Action       string `json:"action"`
	}
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}
	if req.SubmissionID == "" || req.Action == "" || req.ExpectedTurn <= 0 {
		writeError(w, http.StatusBadRequest, "submission_id, action, and a positive expected_turn are required")
		return
	}

	playMu.Lock()
	c, ok := requirePlayCampaign(w, campaignID)
	if !ok {
		return
	}
	if c.Owner != username && !playIsCampaignMember(c, username) {
		playMu.Unlock()
		writeError(w, http.StatusForbidden, "must be a campaign member")
		return
	}

	if c.SafeTurnCurrent == 0 {
		c.SafeTurnCurrent = 1
	}

	if c.SafeTurnSubmissionIDs != nil && c.SafeTurnSubmissionIDs[req.SubmissionID] {
		playMu.Unlock()
		writeError(w, http.StatusConflict, "submission_id already used in this campaign")
		return
	}

	if req.ExpectedTurn != c.SafeTurnCurrent {
		current := c.SafeTurnCurrent
		playMu.Unlock()
		writeJSON(w, http.StatusConflict, map[string]interface{}{"current_turn": current})
		return
	}

	acceptedTurn := c.SafeTurnCurrent
	c.SafeTurnCurrent++
	entry := &playSafeTurn{
		SubmissionID: req.SubmissionID,
		Action:       req.Action,
		AcceptedTurn: acceptedTurn,
		NextTurn:     c.SafeTurnCurrent,
	}
	c.SafeTurns = append(c.SafeTurns, entry)
	if c.SafeTurnSubmissionIDs == nil {
		c.SafeTurnSubmissionIDs = make(map[string]bool)
	}
	c.SafeTurnSubmissionIDs[req.SubmissionID] = true
	resp := map[string]interface{}{
		"submission_id": entry.SubmissionID,
		"action":        entry.Action,
		"accepted_turn": entry.AcceptedTurn,
		"next_turn":     entry.NextTurn,
	}
	playMu.Unlock()
	persistState()

	writeJSON(w, http.StatusCreated, resp)
}

func handleListPlaySafeTurns(w http.ResponseWriter, r *http.Request, campaignID string) {
	username, ok := authenticatePlay(r)
	if !ok {
		writeError(w, http.StatusUnauthorized, "authentication required")
		return
	}

	playMu.Lock()
	c, ok := requirePlayCampaign(w, campaignID)
	if !ok {
		return
	}
	if c.Owner != username && !playIsCampaignMember(c, username) {
		playMu.Unlock()
		writeError(w, http.StatusForbidden, "must be a campaign dm or member")
		return
	}

	current := c.SafeTurnCurrent
	if current == 0 {
		current = 1
	}
	accepted := make([]map[string]interface{}, 0, len(c.SafeTurns))
	for _, t := range c.SafeTurns {
		accepted = append(accepted, map[string]interface{}{
			"submission_id": t.SubmissionID,
			"action":        t.Action,
			"accepted_turn": t.AcceptedTurn,
			"next_turn":     t.NextTurn,
		})
	}
	playMu.Unlock()

	writeJSON(w, http.StatusOK, map[string]interface{}{
		"current_turn": current,
		"accepted":     accepted,
	})
}
